rust/xaynet-core/src/mask/object/serialization/unit.rs

```diff
@@ -0,0 +1,195 @@
+//! Serialization of masked units.
+//!
+//! See the [mask module] documentation since this is a private module anyways.
+//!
+//! [mask module]: ../index.html
+
+use std::ops::Range;
+
+use anyhow::{anyhow, Context};
+use num::bigint::BigUint;
+
+use crate::{
+    mask::{
+        config::{serialization::MASK_CONFIG_BUFFER_LEN, MaskConfig},
+        object::MaskUnit,
+    },
+    message::{
+        traits::{FromBytes, ToBytes},
+        utils::range,
+        DecodeError,
+    },
+};
+
+const MASK_CONFIG_FIELD: Range<usize> = range(0, MASK_CONFIG_BUFFER_LEN);
+
+/// A buffer for serialized mask units.
+pub struct MaskUnitBuffer<T> {
+    inner: T,
+}
+
+impl<T: AsRef<[u8]>> MaskUnitBuffer<T> {
+    pub fn new(bytes: T) -> Result<Self, DecodeError> {
+        let buffer = Self { inner: bytes };
+        buffer
+            .check_buffer_length()
+            .context("not a valid mask unit")?;
+        Ok(buffer)
+    }
+
+    pub fn new_unchecked(bytes: T) -> Self {
+        Self { inner: bytes }
+    }
+
+    pub fn check_buffer_length(&self) -> Result<(), DecodeError> {
+        let len = self.inner.as_ref().len();
+        if len < MASK_CONFIG_FIELD.end {
+            return Err(anyhow!(
+                "invalid buffer length: {} < {}",
+                len,
+                MASK_CONFIG_FIELD.end
+            ));
+        }
+
+        let total_expected_length = self.try_len()?;
+        if len < total_expected_length {
+            return Err(anyhow!(
+                "invalid buffer length: expected {} bytes but buffer has only {} bytes",
+                total_expected_length,
+                len
+            ));
+        }
+        Ok(())
+    }
+
+    pub fn try_len(&self) -> Result<usize, DecodeError> {
+        let config =
+            MaskConfig::from_byte_slice(&self.config()).context("invalid MaskUnit buffer")?;
+        let data_length = config.bytes_per_number();
+        Ok(MASK_CONFIG_FIELD.end + data_length)
+    }
+
+    pub fn len(&self) -> usize {
+        let config = MaskConfig::from_byte_slice(&self.config()).unwrap();
```

Comment:
>i guess this pattern appears in the other buffers in a similar fashion, maybe this should be addressed in a separate refactoring pr.

yes, although maybe not the best convention, it is as least consistent across the buffer types. Agree that it would be convenient to just remove the possibility of constructing an unchecked buffer if it's not really needed - we can tackle this refactoring task separately.